internal/pool/pool.go

```diff
@@ -140,9 +149,16 @@ func (p *ConnPool) addIdleConn() error {
 	}
 
 	p.connsMu.Lock()
+	defer p.connsMu.Unlock()
+
+	// It is not allowed to add new connections to the closed connection pool.
+	if p.closed() {
+		_ = cn.Close()
+		return ErrClosed
+	}
```

Comment:
I believe this is the only change we need. I agree that `runGoroutine` should be removed.